File: src/numeric/max/max.clj
(comment
  ;; Find the maximum value of a set of values
  )

; Finds the maximum value
;   No generics, but this just works like a generic case natively
(defn mymax [values]
  (reduce (fn [curr next]
            (if (> next curr)
              next
              curr)) values))

; Gets the command line arguments as an integer values
(defn argsAsInts [args default]
  (if (>= (count args) 2)
    (map (fn [val] (Integer/valueOf val)) (drop 1 args))
    default))

; Our main run code; pulls the command line arguments OR 15, 10,
; find the max value, and prints them all to the screen.
(let [args *command-line-args*
      intargs (argsAsInts args [15, 10])
      maxval (mymax intargs)]
  (prn "values: ") (println intargs)
  (printf "%d\n" maxval))

File: src/numeric/max/max.jl
#=
    Finds the maximum value from any set of values
=#

"""
Finds the maximum of a set of values

Julia's type system makes this inherently generic by not specifying a type
"""
function max(values::Vector{T}) where T <: Number
    current = zero(T)
    for value in values
        if current < value
            current = value
        end
    end
    return current
end

# We try to grab all the command lines or just use 15, 10
list = [15, 10]
if length(ARGS) > 0
    list = [parse(Int, arg) for arg in ARGS]
end

maxValue = max(list)

println("values: ", list)
println("max: ", maxValue)
